examples/Cassie/osc_jump/jumping_event_based_fsm.cc

```diff
@@ -0,0 +1,187 @@
+using drake::multibody::MultibodyPlant;
+using drake::systems::BasicVector;
+using drake::systems::Context;
+using drake::systems::DiscreteUpdateEvent;
+using drake::systems::DiscreteValues;
+using drake::systems::EventStatus;
+using Eigen::VectorXd;
+using std::string;
+using std::vector;
+
+namespace dairlib {
+namespace examples {
+
+JumpingEventFsm::JumpingEventFsm(const MultibodyPlant<double>& plant,
+                                 const vector<double>& transition_times,
+                                 bool contact_based, double delay_time,
+                                 FSM_STATE init_state, SIMULATOR simulator_type)
+    : plant_(plant),
+      transition_times_(transition_times),
+      contact_based_(contact_based),
+      transition_delay_(delay_time),
+      init_state_(init_state),
+      simulator_type_(simulator_type) {
+  state_port_ =
+      this->DeclareVectorInputPort(OutputVector<double>(plant.num_positions(),
+                                                        plant.num_velocities(),
+                                                        plant.num_actuators()))
+          .get_index();
+
+  // Configure the contact info port for the particular simulator
+  if (simulator_type_ == DRAKE) {
+    contact_port_ = this->DeclareAbstractInputPort(
+                            "lcmt_contact_info",
+                            drake::Value<drake::lcmt_contact_results_for_viz>{})
+                        .get_index();
+  } else if (simulator_type_ == MUJOCO) {
+    contact_port_ = this->DeclareAbstractInputPort(
+                            "lcmt_contact_info",
+                            drake::Value<dairlib::lcmt_contact_mujoco>{})
+                        .get_index();
+  } else if (simulator_type_ == GAZEBO) {
+  }
+  this->DeclareVectorOutputPort(BasicVector<double>(1),
+                                &JumpingEventFsm::CalcFiniteState);
+  DeclarePerStepDiscreteUpdateEvent(&JumpingEventFsm::DiscreteVariableUpdate);
+
+  BasicVector<double> init_prev_time = BasicVector<double>(VectorXd::Zero(1));
+  BasicVector<double> init_state_trigger_time =
+      BasicVector<double>(VectorXd::Zero(1));
+  BasicVector<double> init_fsm_state = BasicVector<double>(VectorXd::Zero(1));
+  init_state_trigger_time.get_mutable_value()(0) = -1.0;
+  init_fsm_state.get_mutable_value()(0) = init_state_;
+
+  prev_time_idx_ = this->DeclareDiscreteState(init_prev_time);
+  guard_trigger_time_idx_ = this->DeclareDiscreteState(init_state_trigger_time);
+  fsm_idx_ = this->DeclareDiscreteState(init_fsm_state);
+  transition_flag_idx_ = this->DeclareDiscreteState(1);
+}
+
+EventStatus JumpingEventFsm::DiscreteVariableUpdate(
+    const Context<double>& context,
+    DiscreteValues<double>* discrete_state) const {
+  // Get inputs to the leaf system
+  const OutputVector<double>* state_feedback =
+      (OutputVector<double>*)this->EvalVectorInput(context, state_port_);
+  const drake::AbstractValue* contact_info =
```

Comment:
There's now a cleaner way to evaluate abstract inputs that skips the later `get_value<T>` line
```
const auto &msg = this->EvalInputValue<dairlib::lcmt_contact_mujoco>(contact_port_)
```

Reply:
Oh good to know! I think in this case it actually makes sense to cast the `AbstractValue*` because the type of contact_info differs between Drake and Mujoco.